DatadogInternal/Sources/DataStore/DataStore.swift

```diff
@@ -0,0 +1,96 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+/// Possible results of retrieving a value from the data store.
+public enum DataStoreValueResult {
+    /// Indicates that there was no value associated with the requested key.
+    case noValue
+    /// Indicates that the value was successfully found and retrieved.
+    /// The retrieved value corresponds to the expected version and is accessible according to its encoded policy.
+    case value(Data)
+    /// Indicates that a value was found, but it was stored under a different version than expected by the caller.
+    /// Upon receiving this result, the value associated with the key is deleted.
+    case incompatibleVersion(actual: DataStoreKeyVersion, expected: DataStoreKeyVersion)
+    /// Indicates that a value was found, but it was stored with a policy that disallows passing it to the caller.
+    /// Upon receiving this result, the value associated with the key is deleted.
+    case incompatiblePolicy(DataStoreKeyPolicy)
+    /// Indicates that the value could not be read due to an underlying error.
+    /// This may represent an error with the data format itself or an I/O exception that occurred during reading.
+    case error(Error)
+
+    /// Retrieves the data associated with the result, if available.
+    public var data: Data? {
+        switch self {
+        case .value(let data): return data
+        case .noValue, .incompatibleVersion, .incompatiblePolicy, .error: return nil
+        }
+    }
+}
+
+/// Defines the interface for a data store capable of storing key-value pairs for a given feature.
+public protocol DataStore {
+    /// Sets the value for the specified key in the data store.
+    ///
+    /// - Parameters:
+    ///   - value: The data to be stored.
+    ///   - key: The unique identifier for the data. Must be a valid file name, as it will be persisted in files.
+    ///   - version: The version associated with the data.
+    ///   - policy: The policy governing access to the stored data.
+    func set(value: Data, forKey key: String, version: DataStoreKeyVersion, policy: DataStoreKeyPolicy)
+
+    /// Retrieves the value associated with the specified key from the data store.
+    ///
+    /// - Parameters:
+    ///   - key: The unique identifier for the data. Must be a valid file name, as it will be persisted in files.
+    ///   - version: The version expected for the retrieved data.
+    ///   - callback: A closure to be executed once the retrieval operation completes, providing the result.
+    func value(forKey key: String, version: DataStoreKeyVersion, callback: @escaping (DataStoreValueResult) -> Void)
```

Comment:
@maciejburda @maxep I'm happy to discuss alternatives to this design, notably for removing `policy` from the "set value" API. The interface proposed in this PR might be a bit confusing for values set in one place and read in another part of the codebase (like "last RUM view" information). The more "set" and "get" are distanced, the harder it is to plan the right storage policies.

Considering policies we introduce in this PR:

`deleteUponRead` could be replaced with either:
-  A return value in "get value" callback - this would let the caller decide if delete or not based on the value that was read. It will make a compromise on atomicity assuming that the caller makes a synchronous side-effect right upon read and app is terminated due to crash (data store may not be given enough time to delete). In practice tho, all writes in our codebase are async, hence the risk of race condition is very low.
- A `delete` parameter in "get value" API - this way the caller can decide upfront if the value should be deleted after read. It will give the same atomicity as current implementation.

`sameAppVersion` and `sameOSVersion` can be replaced with either:
-  The caller encoding both informations as part of the value. That would require the caller to access `DatdogContext` which is not always convenient (e.g. we don't have it in App Hangs observer).
-  By considering them standard attributes of `DataStoreValueResult` (set by data store implementation).

There are perhaps other options that we can think of. I don't mind doing a short 🧠⚡  meeting on this design. LMK.

Reply:
I was writing a comment about that :) The `PolicyData` is too opinionated and add a level of complexity regarding queues sync. IMO, we should let the feature handling its data and constraints: If its data is linked to a app/os version or any other context data, then the caller will have to get that info before writing.
Same goes for the `DataStoreKeyVersion`: If the format has changed, then the feature will fail to decode and that's it, no need to care about it from the core perspective.

I think the policy should be more about rules on how to manage persistency of the data: like `deleteUponRead` that is present here. We could also think of ttl or expiration date policy, tho it should be added only if we have a use case for it.

I'm happy to have brainstorming session on this 👍

Reply:
The PR is now updated with the new discussed design 👍. Thanks for sharing your points @maxep, I think we have something (simpler and) better now 💪. No more `policy` consideration in `DS` implementation.